PWA 12.5 docs (#143)
AdobeDocs/commerce-pwa-studio: src/pages/integrations/analytics/custom-events/index.md

---

> title: Custom events for Beacon analytics

Reviewer: the official name is Experience Platform(I think), not Beacon

---

# Creating custom events

PWA Studio 12.5 contains an eventing framework with a series of default events that cover common scenarios.
By creating and importing a custom extension, it is possible to add new events so that you may track specific user interactions beyond what is included in the default package.

PWA Studio provides a [sample extension](https://github.com/magento/pwa-studio/tree/develop/packages/extensions/experience-platform-connector/) that shows how to add a custom event.

> In the sample extension, the `SEARCH_REPSONSE` event is being tracked. This event is fired when a search result is returned from the Adobe Experience Manager (AEM), with either results or a `EMPTY_RESULT` response.

Reviewer: * I hope it's not spelled `SEARCH_REPSONSE` in the codebase 😬 
* the event is fired when Magento returns search results data, not AEM

These files can be used as a starting point for a custom event extension.

## searchResponseReceived.js

> This file sets the payload and the search results context. The custom event name is also defined in this file.

Reviewer: Seems to be a lot of relevant code missing from this snippet.

Also, I would describe this file as an event handler file that returns two functions:

1. `canHandle()` - a function that receives an event from PWA Studio's eventing framework and returns true if the `handle()` function in this file should handle the incoming event
2. `handle()` - function that handles the event

```javascript
const canHandle = event => event.type === 'SEARCH_RESPONSE';

...

const { payload } = event;

    const {
        categories,
        facets,
        page,
        perPage,
        products,
        searchRequestId,
        searchUnitId,
        suggestions
    } = payload;

    const searchResultsContext = {
        units: [
            {
                categories,
                facets,
                page,
                perPage,
                products,
                searchRequestId,
                searchUnitId,
                suggestions
            }
        ]
    };

    sdk.context.setSearchResults(searchResultsContext);
    sdk.publish.searchResponseReceived(searchUnitId, searchResultsContext);
};
```

## wrapUseAutocomplete.js

> In `wrapUseAutocomplete.js` the search result is received and the payload defined and dispatched. This take place in the `useEffect` function:

Reviewer: Would describe this file as a talon wrapper that wraps around the `useAutocomplete()` talon to add event dispatching logic.

Would link to doc page about [talons](https://developer.adobe.com/commerce/pwa-studio/guides/general-concepts/talons/), tutorial on [how to wrap a talon](https://developer.adobe.com/commerce/pwa-studio/tutorials/targets/modify-talon-results/), and whatever docs we have about the eventing framework.

> ```javascript

Reviewer: Missing a lot of relevant code

Reviewer 2: Deferring to James.

useEffect(() => {
            if (
                messageType === 'RESULT_SUMMARY' ||
                messageType === 'EMPTY_RESULT'
            ) {
                dispatch({
                    type: 'SEARCH_RESPONSE',
                    payload: {
                        categories: categories || [],
                        facets: [],
                        page: 1,
                        perPage: 3, // Same value used in GQL query
                        products: displayResult || [],
                        searchRequestId: value,
                        searchUnitId: 'search-bar',
                        suggestions: displayResult || []
                    }
                });
            }
```

## intercept.js

> Edit copy of `intercept.js` to pull in the custom extension by name. In this case, the `SearchBar` wrapper:

Reviewer: This is an intercept file that interacts with PWA Studio's extensibility framework to modify code during build time.
In this particular code, the `useApp()`, `useAccountMenu()`, and `useAutocomplete()` talons are wrapped with modules that provide additional logic.

Would link to:
* [docs on extensibility](https://developer.adobe.com/commerce/pwa-studio/guides/general-concepts/extensibility/)
* [modify talons tutorial](https://developer.adobe.com/commerce/pwa-studio/tutorials/targets/modify-talon-results/)

```javascript
    talons.tap(({ App, Header, SearchBar }) => {
        App.useApp.wrapWith('@magento/experience-platform-connector');
        Header.useAccountMenu.wrapWith(
            '@magento/experience-platform-connector/src/wrappers/wrapUseAccountMenu'
        );
        SearchBar.useAutocomplete.wrapWith(
            '@magento/experience-platform-connector/src/wrappers/wrapUseAutocomplete'
        );
```

> ## Import a custom extension

Reviewer: Not sure this is needed. Most developers should already know how to set up a new package but if you really want to start from scratch, I would refer to the section in this tutorial on [how to initialize a project](https://developer.adobe.com/commerce/pwa-studio/tutorials/targets/modify-talon-results/#initialize-the-project).

I would replace this particular section with a testing locally section that looks like [this](https://developer.adobe.com/commerce/pwa-studio/tutorials/targets/modify-talon-results/#test-on-a-local-instance)

---
title: Custom events for Experience Platform analytics
---

# Creating custom events

PWA Studio 12.5 contains an eventing framework with a series of default events that cover common scenarios.
By creating and importing a custom extension, it is possible to add new events so that you may track specific user interactions beyond what is included in the default package.

PWA Studio provides a [sample extension](https://github.com/magento/pwa-studio/tree/develop/packages/extensions/experience-platform-connector/) that shows how to add a custom event.

In the sample extension, the `SEARCH_RESPONSE` event is being tracked. This event is fired when a search result is returned from Adobe Commerce, with either results or a `EMPTY_RESULT` response.

> These files can be used as a starting point for a custom event extension.

Reviewer 2: ```suggestion
These files can be used as a starting point to create your own custom events.
```
